File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        left = [0 for _ in range(len(nums))]
        right = [0 for _ in range(len(nums))]

        # for left
        left[0] = 1

        for i in range(1,len(nums)):
            left[i] = nums[i-1] * left[i-1]

        # for right
        right[len(nums) - 1] = 1

        for j in range(len(nums) - 2,-1,-1):
            right[j] = nums[j+1] * right[j+1] 
        
        # for result
        for i in range(len(nums)):
            nums[i] = left[i] * right[i]

        return nums

        # SC -> O(N)
        # TC -> O(N)

        

File: 238-product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        // [1,2,3,4]
        // left = [1,1,2,6]
        // result = [24,12,8,6]

        int[] lNums = new int[nums.length];

        lNums[0] = 1;

        for(int i = 1; i < lNums.length;i++){
            lNums[i] = lNums[i-1] * nums[i-1];
        }

        int rightProd = 1;

        for(int i = lNums.length - 1; i >= 0; i--){
            lNums[i] *= rightProd;
            rightProd *= nums[i];   
        }


        return lNums;        
    }
}

// SC -> O(N)
// TC -> O(N)